app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java

```diff
@@ -184,161 +181,217 @@ public boolean contactExists(final String name) throws IOException {
+        LOGGER.info("Started loading database");
         final GsonBuilder gsonBuilder = new GsonBuilder().registerTypeHierarchyAdapter(IKey.class, new GsonKeyAdapter());
         final Gson gsonParser = gsonBuilder.create();
+        helper.checkPassword(password);
 
-        // Update the database
-        final Writer writer = this.fileSystem.getWriter(this.profilesPath);
-        gsonParser.toJson(profiles, writer);
-        writer.close();
+        try {
+
+            final String databaseFileString = helper.readDatabaseToString();
+            final SecretKey key = EncryptedSaver.makeKey(password);
+            databaseMap.put(ENCRYPTION_KEY, key);
+            final String databaseJson = EncryptedSaver.encryptOrDecryptWithPassword(databaseFileString, key, false);
+            final Database database = gsonParser.fromJson(databaseJson, TYPE_DATABASE);
+            databaseMap.put(DATABASE, database);
+        } catch (InvalidKeySpecException e) {
+            throw new IOException(BAD_KEY_SPEC, e);
+        } catch (IllegalBlockSizeException e) {
+            throw new IOException(DATABASE_WRONG_SIZE, e);
+        } catch (BadPaddingException e) {
+            LOGGER.error(BAD_PADDING, e);
+        }
     }
 
+
     /**
      * {@inheritDoc}
      */
     @Override
-    public void updateProfile(final Profile oldProfile, final Profile newProfile) throws IllegalArgumentException, IOException {
-        // Get the list of contacts
-        final List<Profile> profiles = this.getProfiles();
-        final int lengthBefore = profiles.size();
-        profiles.remove(oldProfile);
-
-        // Throw if the contact to update is not found
-        if (profiles.size() == lengthBefore) {
-            throw new IllegalArgumentException(CONTACT_NOT_FOUND);
-        }
+    public void createDatabase(final Profile profile, final String password) throws IOException {
+        LOGGER.info("Started creating database for profile");
+        initializePassword(password);
+        initializeDatabase(profile, password);
+        LOGGER.info("Database successfully created");
 
-        final GsonBuilder gsonBuilder = new GsonBuilder().registerTypeHierarchyAdapter(IKey.class, new GsonKeyAdapter());
-        final Gson gsonParser = gsonBuilder.create();
+    }
 
-        // Add the new contact and update the database
-        profiles.add(newProfile);
-        final Writer writer = this.fileSystem.getWriter(this.profilesPath);
-        gsonParser.toJson(profiles, writer);
-        writer.close();
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean checkFirstUse() {
+        final File filePass = new File(passwordPath);
+        final File fileDb = new File(databasePath);
+        return filePass.exists() && fileDb.exists();
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public String getUserPassword() throws IOException {
-        return null;
+    public void deleteDatabase() {
```

Comment:
A method like deleteFile(final String path) in AndroidFileSystemAdapter would better suit the intended functionality of this class